File: studypilot_backend/apps/memory/records.py
"""Record format for Exam Mistake Memory.

Walrus Memory is append-only: there is no update or delete. So nothing here
stores a counter. Every event is a new record, and counts are derived by
recalling records and counting them.

The record text *is* the schema. A pipe-delimited header carries the fields we
need to rank on, and the prose underneath exists so semantic recall can find it.
A record without a valid header cannot be parsed later, which is why `parse`
returns None for those rather than guessing: callers count the failures instead
of silently dropping them.

Pure functions only. No I/O, no Django, no network, so this is unit-testable
without credentials or a relayer.
"""
import re
import unicodedata
from dataclasses import dataclass
from datetime import date, timedelta

VALID_KINDS = ("MISS", "HIT", "MASTERED", "PATTERN", "SESSION", "MATERIAL")
SEVERITY_WEIGHTS = {"high": 3, "medium": 2, "low": 1}
DEFAULT_SEVERITY = "medium"
MASTERY_DAYS = 30

# SESSION records carry no topic, so the topic group is optional. The negative
# lookahead stops an absent topic from letting the date slide into the topic
# slot, which would otherwise make every SESSION record parse with topic set to
# a date and no date at all.
_HEADER = re.compile(
    r"^(?P<kind>MISS|HIT|MASTERED|PATTERN|SESSION|MATERIAL)\s*\|\s*"
    r"(?P<namespace>[^|]+?)\s*\|\s*"
    r"(?:(?P<topic>(?!\d{4}-\d{2}-\d{2}\s*(?:\||$))[^|]+?)\s*\|\s*)?"
    r"(?P<date>\d{4}-\d{2}-\d{2})"
    r"(?P<rest>\s*\|.*)?$"
)

_SEVERITY = re.compile(r"sev:\s*(high|medium|low)", re.IGNORECASE)
_EXPIRES = re.compile(r"expires:\s*(\d{4}-\d{2}-\d{2})")
_MISCONCEPTION = re.compile(r"My misconception:\s*(.+?)(?:\n|$)", re.IGNORECASE | re.DOTALL)
_SOURCE = re.compile(r"source:\s*([a-z0-9_]+)", re.IGNORECASE)


@dataclass(frozen=True)
class ParsedRecord:
    kind: str
    namespace: str
    topic: str
    on: date
    severity: str
    expires: date
    text: str


def slugify_topic(value):
    """Stable join key across every record type.

    Drift here is the one failure mode that quietly destroys the whole record:
    "beta blocker selectivity" and "Beta-Blocker Selectivity" must collapse to
    the same slug or their misses will never be counted together.
    """
    if not value:
        return ""
    text = unicodedata.normalize("NFKD", str(value))
    text = text.encode("ascii", "ignore").decode("ascii").lower()
    text = re.sub(r"[^a-z0-9]+", "-", text)
    return text.strip("-")[:120]


def _on(on):
    return (on or date.today()).isoformat()


def _severity(severity):
    value = (severity or DEFAULT_SEVERITY).lower()
    return value if value in SEVERITY_WEIGHTS else DEFAULT_SEVERITY


def build_miss(namespace, topic, question, answered, misconception, correct, severity=DEFAULT_SEVERITY, on=None):
    return (
        f"MISS | {namespace} | {slugify_topic(topic)} | {_on(on)} | sev:{_severity(severity)}\n"
        f"Q: {question}\n"
        f"I answered: {answered}. My misconception: {misconception}.\n"
        f"Correct: {correct}"
    )


def build_hit(namespace, topic, on=None):
    return (
        f"HIT | {namespace} | {slugify_topic(topic)} | {_on(on)}\n"
        f"Answered correctly on a previously missed topic."
    )


def build_mastered(namespace, topic, correct_dates, on=None):
    when = on or date.today()
    expires = when + timedelta(days=MASTERY_DAYS)
    dates = ", ".join(d.isoformat() if hasattr(d, "isoformat") else str(d) for d in correct_dates)
    return (
        f"MASTERED | {namespace} | {slugify_topic(topic)} | {when.isoformat()} | expires:{expires.isoformat()}\n"
        f"Correct on {dates}. Spot-check again after the expiry date."
    )


def build_material(namespace, topic, source_type, title, summary, reference="", on=None):
    """What the student actually studied, as opposed to what they got wrong.

    Kept in its own namespace so it never dilutes the mistake recall the
    weakness briefing ranks over.
    """
    lines = [
        f"MATERIAL | {namespace} | {slugify_topic(topic)} | {_on(on)} | source:{source_type}",
        f"Studied: {title}",
    ]
    if summary:
        lines.append(f"Covers: {summary}")
    if reference:
        lines.append(f"Reference: {reference}")
    return "\n".join(lines)


def source_of(text):
    """The source tag from a MATERIAL header, for example pdf or youtube."""
    match = _SOURCE.search(str(text or ""))
    return match.group(1).strip() if match else ""


def build_pattern(pattern_name, body, subjects, on=None):
    tags = ",".join(subjects or [])
    return (
        f"PATTERN | exam-intel | {slugify_topic(pattern_name)} | {_on(on)} | subjects:{tags}\n"
        f"{body}"
    )


def build_session(namespace, topics, drilled, new_misses, hits, on=None):
    slugs = ", ".join(slugify_topic(topic) for topic in topics if slugify_topic(topic))
    return (
        f"SESSION | {namespace} | {_on(on)} | drilled:{drilled} new_misses:{new_misses} hits:{hits}\n"
        f"Topics: {slugs}"
    )


def misconception_of(text):
    """Pull the stored misconception out of a MISS record, verbatim.

    Quoting it back unchanged is the point: paraphrasing it into something
    softer loses the exact wrong model the student had.
    """
    match = _MISCONCEPTION.search(str(text or ""))
    return match.group(1).strip().rstrip(".") if match else ""


def parse(text):
    """Parse one stored record, or None if it carries no valid header.

    `memwal_analyze` writes in its own words and will land here headerless.
    Returning None rather than a best guess keeps unparseable memories countable,
    so formatting drift shows up as a number instead of a briefing that quietly
    degrades.
    """
    if not text:
        return None
    lines = str(text).strip().splitlines()
    if not lines:
        return None
    header = _HEADER.match(lines[0].strip())
    if not header:
        return None

    try:
        on = date.fromisoformat(header.group("date"))
    except ValueError:
        return None

    rest = header.group("rest") or ""
    severity_match = _SEVERITY.search(rest)
    expires_match = _EXPIRES.search(rest)
    try:
        expires = date.fromisoformat(expires_match.group(1)) if expires_match else None
    except ValueError:
        expires = None

    return ParsedRecord(
        kind=header.group("kind"),
        namespace=(header.group("namespace") or "").strip(),
        topic=slugify_topic(header.group("topic")),
        on=on,
        severity=severity_match.group(1).lower() if severity_match else DEFAULT_SEVERITY,
        expires=expires,
        text=str(text),
    )

File: studypilot_worker/src/memory/records.ts
// Record format for Exam Mistake Memory. A port of apps/memory/records.py that
// writes byte-identical records, so memory written by the Django backend and by
// this Worker is one history, not two.
//
// Walrus Memory is append-only: there is no update or delete. So nothing here
// stores a counter. Every event is a new record, and counts are derived by
// recalling records and counting them.
//
// The record text *is* the schema. A pipe-delimited header carries the fields
// we rank on, and the prose underneath exists so semantic recall can find it.
// Pure functions only: no I/O, so this is testable without credentials.

export const SEVERITY_WEIGHTS: Record<string, number> = { high: 3, medium: 2, low: 1 };
export const DEFAULT_SEVERITY = "medium";
export const MASTERY_DAYS = 30;

// SESSION records carry no topic, so the topic group is optional. The negative
// lookahead stops an absent topic from letting the date slide into the topic
// slot, which would otherwise parse every SESSION with topic set to a date.
const HEADER =
  /^(?<kind>MISS|HIT|MASTERED|PATTERN|SESSION|MATERIAL|PROGRESS)\s*\|\s*(?<namespace>[^|]+?)\s*\|\s*(?:(?<topic>(?!\d{4}-\d{2}-\d{2}\s*(?:\||$))[^|]+?)\s*\|\s*)?(?<date>\d{4}-\d{2}-\d{2})(?<rest>\s*\|.*)?$/;

const SEVERITY = /sev:\s*(high|medium|low)/i;
const EXPIRES = /expires:\s*(\d{4}-\d{2}-\d{2})/;
const MISCONCEPTION = /My misconception:\s*([\s\S]+?)(?:\n|$)/i;
const SOURCE = /source:\s*([a-z0-9_]+)/i;
const MINUTES = /minutes:\s*(\d+)/i;
const STATUS = /status:\s*(active|done)/i;
const AT = /at:\s*(\S+)/;

export interface ParsedRecord {
  kind: string;
  namespace: string;
  topic: string;
  on: string; // YYYY-MM-DD
  severity: string;
  expires: string | null;
  text: string;
}

/** Today's date in UTC, matching Django's TIME_ZONE = "UTC". */
export function today(): string {
  return new Date().toISOString().slice(0, 10);
}

export function addDays(isoDate: string, days: number): string {
  const d = new Date(`${isoDate}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

export function daysBetween(from: string, to: string): number {
  return Math.round((Date.parse(`${to}T00:00:00Z`) - Date.parse(`${from}T00:00:00Z`)) / 86400000);
}

function validDate(value: string): boolean {
  const d = new Date(`${value}T00:00:00Z`);
  return !Number.isNaN(d.getTime()) && d.toISOString().slice(0, 10) === value;
}

/**
 * Stable join key across every record type. "beta blocker selectivity" and
 * "Beta-Blocker Selectivity" must collapse to the same slug or their misses
 * will never be counted together.
 */
export function slugifyTopic(value: unknown): string {
  if (!value) return "";
  // NFKD then drop non-ASCII, like Python's encode("ascii", "ignore").
  const ascii = String(value).normalize("NFKD").replace(/[^\x00-\x7f]/g, "").toLowerCase();
  return ascii.replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "").slice(0, 120);
}

function on(date?: string): string {
  return date || today();
}

function severityOf(severity?: string): string {
  const value = (severity || DEFAULT_SEVERITY).toLowerCase();
  return value in SEVERITY_WEIGHTS ? value : DEFAULT_SEVERITY;
}

export function buildMiss(
  namespace: string,
  topic: string,
  question: string,
  answered: string,
  misconception: string,
  correct: string,
  severity = DEFAULT_SEVERITY,
  date?: string,
): string {
  return (
    `MISS | ${namespace} | ${slugifyTopic(topic)} | ${on(date)} | sev:${severityOf(severity)}\n` +
    `Q: ${question}\n` +
    `I answered: ${answered}. My misconception: ${misconception}.\n` +
    `Correct: ${correct}`
  );
}

export function buildHit(namespace: string, topic: string, date?: string): string {
  return `HIT | ${namespace} | ${slugifyTopic(topic)} | ${on(date)}\nAnswered correctly on a previously missed topic.`;
}

export function buildMastered(namespace: string, topic: string, correctDates: string[], date?: string): string {
  const when = on(date);
  const expires = addDays(when, MASTERY_DAYS);
  return (
    `MASTERED | ${namespace} | ${slugifyTopic(topic)} | ${when} | expires:${expires}\n` +
    `Correct on ${correctDates.join(", ")}. Spot-check again after the expiry date.`
  );
}

/**
 * What the student actually studied, as opposed to what they got wrong. Kept in
 * its own namespace so it never dilutes the mistake recall.
 */
export function buildMaterial(
  namespace: string,
  topic: string,
  sourceType: string,
  title: string,
  summary: string,
  reference = "",
  date?: string,
): string {
  const lines = [`MATERIAL | ${namespace} | ${slugifyTopic(topic)} | ${on(date)} | source:${sourceType}`, `Studied: ${title}`];
  if (summary) lines.push(`Covers: ${summary}`);
  if (reference) lines.push(`Reference: ${reference}`);
  return lines.join("\n");
}

export function sourceOf(text: unknown): string {
  const match = SOURCE.exec(String(text ?? ""));
  return match ? match[1].trim() : "";
}

export function buildSession(
  namespace: string,
  topics: string[],
  drilled: number,
  newMisses: number,
  hits: number,
  minutes = 0,
  date?: string,
): string {
  const slugs = topics.map(slugifyTopic).filter(Boolean).join(", ");
  return (
    `SESSION | ${namespace} | ${on(date)} | drilled:${drilled} new_misses:${newMisses} hits:${hits} minutes:${minutes}\n` +
    `Topics: ${slugs}`
  );
}

export function minutesOf(text: unknown): number {
  const match = MINUTES.exec(String(text ?? ""));
  return match ? Number.parseInt(match[1], 10) : 0;
}

/** The stored misconception, verbatim. Paraphrasing it loses the exact wrong model. */
export function misconceptionOf(text: unknown): string {
  const match = MISCONCEPTION.exec(String(text ?? ""));
  return match ? match[1].trim().replace(/\.+$/, "") : "";
}

/**
 * Parse one stored record, or null if it has no valid header. Returning null
 * rather than a best guess keeps unparseable memories countable, so format
 * drift shows up as a number instead of a quietly degrading briefing.
 */
export function parse(text: unknown): ParsedRecord | null {
  if (!text) return null;
  const lines = String(text).trim().split(/\r?\n/);
  if (!lines.length || !lines[0]) return null;
  const header = HEADER.exec(lines[0].trim());
  if (!header?.groups) return null;
  const date = header.groups.date;
  if (!validDate(date)) return null;

  const rest = header.groups.rest || "";
  const severity = SEVERITY.exec(rest);
  const expiresMatch = EXPIRES.exec(rest);
  const expires = expiresMatch && validDate(expiresMatch[1]) ? expiresMatch[1] : null;

  return {
    kind: header.groups.kind,
    namespace: (header.groups.namespace || "").trim(),
    topic: slugifyTopic(header.groups.topic),
    on: date,
    severity: severity ? severity[1].toLowerCase() : DEFAULT_SEVERITY,
    expires,
    text: String(text),
  };
}

/** Python's datetime.isoformat(timespec="microseconds") for a UTC instant. */
export function microsecondStamp(when = new Date(), extraMicros = 0): string {
  const iso = when.toISOString(); // 2026-09-19T10:11:12.345Z
  const micros = String(when.getUTCMilliseconds() * 1000 + extraMicros).padStart(6, "0");
  return `${iso.slice(0, 19)}.${micros}+00:00`;
}

// Workers freeze the clock during a request, so two checkpoints written in one
// request would get the same stamp. A per-isolate counter keeps them ordered.
let lastStamp = "";
let bump = 0;

function nextStamp(): string {
  const base = microsecondStamp(new Date());
  if (base.slice(0, 23) === lastStamp.slice(0, 23)) {
    bump = Math.min(bump + 1, 999);
  } else {
    bump = 0;
  }
  lastStamp = microsecondStamp(new Date(), bump);
  return lastStamp;
}

/**
 * Unfinished work, so it can be picked up on another day or device. The store
 * has no update, so finishing appends a `status:done` record. `at:` carries a
 * microsecond timestamp because several checkpoints of one activity land on
 * the same date and the date alone cannot say which is current.
 */
export function buildProgress(
  namespace: string,
  activityKey: string,
  label: string,
  payload: unknown,
  status = "active",
  at?: string,
  date?: string,
): string {
  const stamp = at || nextStamp();
  return (
    `PROGRESS | ${namespace} | ${slugifyTopic(activityKey)} | ${on(date)} | status:${status} at:${stamp}\n` +
    `Label: ${label}\n` +
    `Payload: ${asciiJson(payload)}`
  );
}

/** json.dumps(..., separators=(",", ":")) with Python's default ensure_ascii. */
export function asciiJson(value: unknown): string {
  return JSON.stringify(value).replace(/[-￿]/g, (ch) => `\\u${ch.charCodeAt(0).toString(16).padStart(4, "0")}`);
}

export function statusOf(text: unknown): string {
  const match = STATUS.exec(String(text ?? ""));
  return match ? match[1].toLowerCase() : "";
}

export function checkpointAt(text: unknown): string {
  const match = AT.exec(String(text ?? ""));
  return match ? match[1] : "";
}

export function payloadOf(text: unknown): unknown {
  for (const line of String(text ?? "").split(/\r?\n/)) {
    if (line.startsWith("Payload: ")) {
      try {
        return JSON.parse(line.slice("Payload: ".length));
      } catch {
        return {};
      }
    }
  }
  return {};
}

export function labelOf(text: unknown): string {
  for (const line of String(text ?? "").split(/\r?\n/)) {
    if (line.startsWith("Label: ")) return line.slice("Label: ".length).trim();
  }
  return "";
}
